syncmaster/backend/providers/auth/keycloak_provider.py

```diff
@@ -0,0 +1,146 @@
+        app.dependency_overrides[AuthProvider] = cls
+        app.dependency_overrides[KeycloakAuthProviderSettings] = lambda: settings
+        return app
+
+    async def get_token_password_grant(
+        self,
+        grant_type: str | None = None,
+        login: str | None = None,
+        password: str | None = None,
+        scopes: list[str] | None = None,
+        client_id: str | None = None,
+        client_secret: str | None = None,
+    ) -> dict[str, Any]:
+        raise NotImplementedError("Password grant is not supported by KeycloakAuthProvider.")
+
+    async def get_token_authorization_code_grant(
+        self,
+        code: str,
+        redirect_uri: str,
+        scopes: list[str] | None = None,
+        client_id: str | None = None,
+        client_secret: str | None = None,
+    ) -> dict[str, Any]:
+        try:
+            redirect_uri = redirect_uri or self.settings.redirect_uri
+            token = self.keycloak_openid.token(
+                grant_type="authorization_code",
+                code=code,
+                redirect_uri=redirect_uri,
+            )
+            return token
+        except Exception as e:
+            raise AuthorizationError("Failed to get token") from e
+
+    async def get_current_user(self, access_token: str, *args, **kwargs) -> Any:
+        request: Request = kwargs["request"]
+        refresh_token = request.session.get("refresh_token")
+
+        if not access_token:
+            log.debug("No access token found in session.")
+            self.redirect_to_auth(request.url.path)
+
+        try:
+            token_info = self.keycloak_openid.decode_token(token=access_token)
+        except Exception as e:
+            log.info("Access token is invalid or expired: %s", e)
+            token_info = None
+
+        if not token_info and refresh_token:
+            log.debug("Access token invalid. Attempting to refresh.")
+
+            try:
+                new_tokens = await self.refresh_access_token(refresh_token)
+
+                new_access_token = new_tokens.get("access_token")
+                new_refresh_token = new_tokens.get("refresh_token")
+                request.session["access_token"] = new_access_token
+                request.session["refresh_token"] = new_refresh_token
+
+                token_info = self.keycloak_openid.decode_token(
+                    token=new_access_token,
+                )
+                log.debug("Access token refreshed and decoded successfully.")
+            except Exception as e:
+                log.debug("Failed to refresh access token: %s", e)
+                self.redirect_to_auth(request.url.path)
+
+        user_id = token_info.get("sub")
+        login = token_info.get("preferred_username")
+        email = token_info.get("email")
+        first_name = token_info.get("given_name")
+        middle_name = token_info.get("middle_name")
+        last_name = token_info.get("family_name")
```

Comment:
Are these fields hardcoded in Keycloak, or they can vary along different Keycloak instances?

Reply:
hardcoded, they are both the same as in open source keycloak and mts keycloak

Reply:
Hardcoded in Keycloak's source code, or this is just a coincidence?

Reply:
[as I understood from source code](https://github.com/keycloak/keycloak/blob/3ca3a4ad349b4d457f6829eaf2ae05f1e01408be/core/src/main/java/org/keycloak/representations/IDToken.java#L39C32-L39C50), hardcoded

Reply:
Ok, please add this link as a comment above these field names